crates/network/src/ethereum/signer.rs

```diff
@@ -0,0 +1,152 @@
+use std::sync::Arc;
+
+/// A signer capable of signing any transaction for the Ethereum network.
+#[derive(Clone)]
+pub struct EthereumSigner(Arc<dyn TxSigner<Signature> + Send + Sync>);
+
+impl std::fmt::Debug for EthereumSigner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("EthereumSigner").finish()
+    }
+}
+
+impl<S> From<S> for EthereumSigner
+where
+    S: TxSigner<Signature> + Send + Sync + 'static,
+{
+    fn from(signer: S) -> Self {
+        Self::new(signer)
+    }
+}
+
+impl EthereumSigner {
+    /// Create a new Ethereum signer.
+    pub fn new<S>(signer: S) -> Self
+    where
+        S: TxSigner<Signature> + Send + Sync + 'static,
+    {
+        Self(Arc::new(signer))
+    }
+
+    async fn sign_transaction(
+        &self,
+        tx: &mut dyn SignableTransaction<Signature>,
+    ) -> alloy_signer::Result<Signature> {
+        self.0.sign_transaction(tx).await
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl NetworkSigner<Ethereum> for EthereumSigner {
+    async fn sign_transaction(&self, tx: TypedTransaction) -> alloy_signer::Result<TxEnvelope> {
+        match tx {
+            TypedTransaction::Legacy(mut t) => {
+                let sig = self.sign_transaction(&mut t).await?;
+                Ok(t.into_signed(sig).into())
+            }
+            TypedTransaction::Eip2930(mut t) => {
+                let sig = self.sign_transaction(&mut t).await?;
+                Ok(t.into_signed(sig).into())
+            }
+            TypedTransaction::Eip1559(mut t) => {
+                let sig = self.sign_transaction(&mut t).await?;
+                Ok(t.into_signed(sig).into())
+            }
+            TypedTransaction::Eip4844(mut t) => {
+                let sig = self.sign_transaction(&mut t).await?;
+                Ok(t.into_signed(sig).into())
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{TxSigner, TxSignerSync};
+    use alloy_consensus::{SignableTransaction, TxLegacy};
+    use alloy_primitives::{address, ChainId, Signature, U256};
+    use alloy_signer::{k256, Result, Signer};
+
+    #[tokio::test]
+    async fn signs_tx() {
+        async fn sign_tx_test(tx: &mut TxLegacy, chain_id: Option<ChainId>) -> Result<Signature> {
+            let mut before = tx.clone();
+            let sig = sign_dyn_tx_test(tx, chain_id).await?;
+            if let Some(chain_id) = chain_id {
+                assert_eq!(tx.chain_id, Some(chain_id), "chain ID was not set");
+                before.chain_id = Some(chain_id);
+            }
+            assert_eq!(*tx, before);
+            Ok(sig)
+        }
+
+        async fn sign_dyn_tx_test(
+            tx: &mut dyn SignableTransaction<Signature>,
+            chain_id: Option<ChainId>,
+        ) -> Result<Signature> {
+            let mut wallet: alloy_signer::Wallet<k256::ecdsa::SigningKey> =
```

Comment:
This generic qualification required? Or can we default to it?

Reply:
it's LocalWallet